polar-core/src/kb.rs

```diff
@@ -75,11 +81,247 @@ impl KnowledgeBase {
+                                break;
+                            }
+                        }
+                        if !found_match {
+                            let err: PolarError = error::RuntimeError::FileLoading {
+                                msg: format!("Rule {}({:#?}) not allowed because parameters did not match rule template.", r.name, r.params),
+                            }
+                            .into();
+                            return Err(err.set_context(None, Some(&r.body)));
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn param_fields_match(&self, template_fields: &Dictionary, rule_fields: &Dictionary) -> bool {
+        return template_fields
+            .fields
+            .iter()
+            .map(|(k, template_value)| {
+                rule_fields
+                    .fields
+                    .get(k)
+                    .map(|rule_value| rule_value == template_value)
+                    .unwrap_or_else(|| false)
+            })
+            .all(|v| v);
+    }
+
+    fn rule_params_match(&self, rule: &Rule, template: &Rule) -> bool {
+        for (rule_param, template_param) in rule.params.iter().zip(template.params.iter()) {
+            let res = match (
+                template_param.parameter.value(),
+                template_param.specializer.as_ref().map(Term::value),
+                rule_param.parameter.value(),
+                rule_param.specializer.as_ref().map(Term::value),
+            ) {
+                (
+                    Value::Variable(_),
+                    Some(Value::Pattern(template_spec)),
+                    Value::Variable(_),
+                    Some(Value::Pattern(rule_spec)),
+                ) => {
+                    // If template and rule both have pattern specializers, then rule
+                    // specializer must match template specializer
+                    match (template_spec, rule_spec) {
+                        (
+                            Pattern::Instance(template_instance),
+                            Pattern::Instance(rule_instance),
+                        ) => {
+                            // if tags match, all template fields must match those in rule fields, otherwise false
+                            if template_instance.tag == rule_instance.tag {
+                                self.param_fields_match(
+                                    &template_instance.fields,
+                                    &rule_instance.fields,
+                                )
+                            // If tags don't match, then rule specializer must be a subclass of template specializer
+                            } else if let Some(template_class) =
+                                self.constants.get(&template_instance.tag)
+                            {
+                                if let Some(rule_mro) = self.mro.get(&rule_instance.tag) {
```

Comment:
Would it be easier for mro to just contain class names?

Reply:
I like your suggestion of the MRO containing instance IDs.